README.md
# iron
Serialize and encrypt any Elixir `term()` and send it around.

"iron is a cryptographic utility for sealing a JSON object using symmetric key encryption with message integrity verification. Or in other words, it lets you encrypt an object, send it around (in cookies, authentication credentials, etc.), then receive it back and decrypt it. The algorithm ensures that the message was not tampered with, and also provides a simple mechanism for password rotation." - [iron](https://github.com/hueniverse/iron)


## Examples

```elixir
    iex> sealed = Iron.seal(%{"a" => 1, "b" => 2, "c" => [3, 4, 5], "d" => %{ "e" => "f" }}, "some_not_random_password_that_is_also_long_enough")
    "Fe26.2**CECQFMECC7VPJBP4IJFKTSG1VTNM41Q3G3I7H195218HV0AHP5A0====*RhZkdeZnjMi2AL0oB7L_Ww*v2BY2UeJknYojSbR5wmHQf-nF9UKcTCWtvL28-N2-9WePDiHgN2MUopWq1vazxv4*1531861757834*OL8TRM1NUPM3COATCES9GHG7UBV4FDONL8P6OKH5FHK0E7P7LS80====*k3TdLkCXsDLHMm6ThVBtmbLZJZUNAhzJEsNxg5D6cIk"

    iex> Iron.unseal(sealed, "some_not_random_password_that_is_also_long_enough")
    %{"a" => 1, "b" => 2, "c" => [3, 4, 5], "d" => %{ "e" => "f" }}
```

## Documentation

[hex documentation for iron](https://hexdocs.pm/iron)


## Installation

```elixir
def deps do
  [{:iron, "~> 0.1.0"}]
end
```

## Acknowledgement

This is an Elixr implementation of the excellent [iron](https://github.com/hueniverse/iron).

## LICENSE

(The MIT License)

Copyright (c) 2018 Benjamin Schultzer

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS,' WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
## IRON v0.1.0 Auguest 22, 2018

### Enhancements

* Initial


test/iron_test.exs
defmodule IronTest do
  use ExUnit.Case
  # doctest Iron

  setup do
    [password: "some_not_random_password_that_is_also_long_enough", map: %{a: 1, b: 2, c: [3, 4, 5], d: %{e: "f" }}]
  end

  test "turns map into a ticket than parses the ticket successfully", %{password: password, map: map} do
    sealed = Iron.seal(map, password)
    unseal = Iron.unseal(sealed, password)
    assert unseal == map
  end

  test "unseal and sealed map with expiration", %{password: password, map: map} do
    sealed = Iron.seal(map, password, %{Iron.defaults() | ttl: 200})
    # passowrd = %{"default" => password}
    unseal = Iron.unseal(sealed, password)
    assert unseal == map
  end

  test "unseal and sealed map with expiration and time offset", %{password: password, map: map} do
    sealed = Iron.seal(map, password, %{Iron.defaults() | ttl: 200, localtime_offset_msec: -10_000})
    # passowrd = %{"default" => password}
    unseal = Iron.unseal(sealed, password, %{Iron.defaults() | localtime_offset_msec: -10_000})
    assert unseal == map
  end

  test "turns map into a ticket than parses the ticket successfully binary", %{map: map} do
    key = Kryptiles.random_bits(256)
    sealed = Iron.seal(map, key)
    unseal = Iron.unseal(sealed, key)
    assert unseal == map
  end

  test "fails to turns map into a ticket password too short", %{password: password, map: map}  do
    key = Kryptiles.random_bits(128)
    sealed = Iron.seal(map, password)
    assert Iron.unseal(sealed, key) == {:error, {500, "Key buffer (password) too small"}}
  end

  # test "fails to turn map into a ticket (failed to stringify map)" do
  #   key = Kryptiles.random_bits(128)
  #   assert Iron.seal([], key) == {:error, "Failed to stringify object"}
  # end

  test "turns map into a ticket than parses the ticket successfully (password map)", %{password: password, map: map} do
    passoword = %{id: "1", secret: password}
    sealed = Iron.seal(map, passoword)
    unseal = Iron.unseal(sealed, passoword)
    assert unseal == map
  end

  test "handles separate password map", %{map: map} do
    key = %{id: "1", encryption: Kryptiles.random_bits(256), integrity: Kryptiles.random_bits(256)}
    sealed = Iron.seal(map, key)
    unseal = Iron.unseal(sealed, key)
    assert unseal == map
  end

  test "handles a common password map", %{map: map} do
    key = %{id: "1", secret: Kryptiles.random_bits(256)}
    sealed = Iron.seal(map, key)
    unseal = Iron.unseal(sealed, key)
    assert unseal == map
  end

  test "fails to parse a sealed object when password not found", %{password: password, map: map} do
    sealed = Iron.seal(map, %{id: "1", secret: password})
    assert Iron.unseal(sealed, %{id: "2", secret: password}) == {:error, {500, "Cannot find password: 1"}}
  end

  describe "generate_key/2" do
    test "returns an error when password is invalid" do
      assert Iron.generate_key(nil, nil) == {:error, {500, "Invalid password nil"}}
    end

    test "returns an error when password is empty" do
      assert Iron.generate_key(<<>>, nil) == {:error, {500, "Empty password"}}
      assert Iron.generate_key([], nil) == {:error, {500, "Empty password"}}
    end

    test "returns an error when password is too short" do
      assert Iron.generate_key('password', Iron.defaults.encryption) == {:error, {500, "Password too short (min 32 characters required)"}}
      assert Iron.generate_key("password", Iron.defaults.encryption) == {:error, {500, "Password string too short (min 32 characters required)"}}
    end

    test "returns an error when options are missing", %{password: password} do
      assert Iron.generate_key(password, nil) == {:error, {500, "Bad options"}}
    end

    test "returns an error when an unknown algorithm is specified", %{password: password} do
      assert Iron.generate_key(password, %{Iron.defaults.encryption | algorithm: "unknown"}) == {:error, {500, "Unknown algorithm: \"unknown\""}}
    end

    test "returns an error when no salt or salt bits are provided", %{password: password} do
      assert Iron.generate_key(password, Map.delete(Iron.defaults.encryption, :salt_bits)) == {:error, {500, "Missing salt or salt_bits options"}}
    end

    test "returns an error when invalid salt bits are provided", %{password: password} do
      assert Iron.generate_key(password, %{Iron.defaults.integrity | iterations: 2, salt_bits: 99999999999999999999}) == {:error, "failed generating random bits"}
    end

    test "returns an error when Kryptiles.random_bits/1 fails", %{password: password} do
      options = Map.put(%{Iron.defaults.encryption | iv_bits:  -1}, :salt, "abcdefg")
      assert Iron.generate_key(password, options) == {:error, "invalid random bits count"}
    end
  end

  describe "encrypt/2" do
    test "returns an error when password is missing" do
      assert Iron.encrypt(<<>>, Iron.defaults.encryption, "data") == {:error, {500, "Empty password"}}
      assert Iron.encrypt([], Iron.defaults.encryption, "data") == {:error, {500, "Empty password"}}
    end
  end

  describe "decrypt/2" do
    test "returns an error when password is missing" do
      assert Iron.decrypt(<<>>, Iron.defaults.encryption, "data") == {:error, {500, "Empty password"}}
      assert Iron.decrypt([], Iron.defaults.encryption, "data") == {:error, {500, "Empty password"}}
    end
  end

  describe "hmac/2" do
    test "returns an error when password is missing" do
      assert Iron.hmac(<<>>, Iron.defaults.integrity, "data") == {:error, {500, "Empty password"}}
      assert Iron.hmac([], Iron.defaults.integrity, "data") == {:error, {500, "Empty password"}}
    end

    test "produces the same mac when used with binary password" do
      data = "not so random"
      key = Kryptiles.random_bits(256)
      options = Iron.defaults.integrity
      mac = Iron.hmac(key, options, data)
      digest = :crypto.hmac(options.algorithm, key, data) |>  Base.encode64() |> String.replace(~r/\+/, "-") |> String.replace(~r/\//, "_") |> String.replace(~r/\=/, "")
      assert digest == mac.digest
    end
  end

  describe "seal/2" do
    test "returns an error when password is invalid" do
      assert Iron.seal("data", nil) == {:error, {500, "Invalid password nil"}}
    end

    test "returns an error when password is empty" do
      assert Iron.seal("data", []) == {:error, {500, "Empty password"}}
      assert Iron.seal("data", <<>>) == {:error, {500, "Empty password"}}
    end

    test "returns an error when integrity options are missing", %{password: password} do
      assert Iron.seal("data", password, %{encryption: %{algorithm: "x"}}) == {:error, {500, "Unknown algorithm: \"x\""}}
    end

    test "returns an error when password.id is invalid" do
      assert Iron.seal("data", %{id: "asd$", secret: "asd"}) == {:error, {500, "Invalid password id: \"asd$\""}}
    end
  end

  describe "unseal/2" do
    test "unseals a ticket", %{password: password, map: map} do
      ticket = "Fe26.2**0cdd607945dd1dffb7da0b0bf5f1a7daa6218cbae14cac51dcbd91fb077aeb5b*aOZLCKLhCt0D5IU1qLTtYw*g0ilNDlQ3TsdFUqJCqAm9iL7Wa60H7eYcHL_5oP136TOJREkS3BzheDC1dlxz5oJ**05b8943049af490e913bbc3a2485bee2aaf7b823f4c41d0ff0b7c168371a3772*R8yscVdTBRMdsoVbdDiFmUL8zb-c3PQLGJn4Y8C-AqI"
      unsealed = Iron.unseal(ticket, password)
      assert unsealed == map
    end

    test "returns an error when number of sealed components is wrong", %{password: password} do
      ticket = "x*Fe26.2**a6dc6339e5ea5dfe7a135631cf3b7dcf47ea38246369d45767c928ea81781694*D3DLEoi-Hn3c972TPpZXqw*mCBhmhHhRKk9KtBjwu3h-1lx1MHKkgloQPKRkQZxpnDwYnFkb3RqdVTQRcuhGf4M**ff2bf988aa0edf2b34c02d220a45c4a3c572dac6b995771ed20de58da919bfa5*HfWzyJlz_UP9odmXvUaVK1TtdDuOCaezr-TAg2GjBCU"
      assert Iron.unseal(ticket, password) == {:error, {500, "Incorrect number of sealed components"}}
    end

    test "returns an error when password is invalid" do
      ticket = "Fe26.2**a6dc6339e5ea5dfe7a135631cf3b7dcf47ea38246369d45767c928ea81781694*D3DLEoi-Hn3c972TPpZXqw*mCBhmhHhRKk9KtBjwu3h-1lx1MHKkgloQPKRkQZxpnDwYnFkb3RqdVTQRcuhGf4M**ff2bf988aa0edf2b34c02d220a45c4a3c572dac6b995771ed20de58da919bfa5*HfWzyJlz_UP9odmXvUaVK1TtdDuOCaezr-TAg2GjBCU"
      assert Iron.unseal(ticket, nil) == {:error, {500, "Invalid password nil"}}
    end

    test "returns an error when password is missing" do
      ticket = "Fe26.2**a6dc6339e5ea5dfe7a135631cf3b7dcf47ea38246369d45767c928ea81781694*D3DLEoi-Hn3c972TPpZXqw*mCBhmhHhRKk9KtBjwu3h-1lx1MHKkgloQPKRkQZxpnDwYnFkb3RqdVTQRcuhGf4M**ff2bf988aa0edf2b34c02d220a45c4a3c572dac6b995771ed20de58da919bfa5*HfWzyJlz_UP9odmXvUaVK1TtdDuOCaezr-TAg2GjBCU"
      assert Iron.unseal(ticket, <<>>) == {:error, {500, "Empty password"}}
      assert Iron.unseal(ticket, []) == {:error, {500, "Empty password"}}
    end

    test "returns an error when mac prefix is wrong", %{password: password} do
      ticket = "Fe27.2**a6dc6339e5ea5dfe7a135631cf3b7dcf47ea38246369d45767c928ea81781694*D3DLEoi-Hn3c972TPpZXqw*mCBhmhHhRKk9KtBjwu3h-1lx1MHKkgloQPKRkQZxpnDwYnFkb3RqdVTQRcuhGf4M**ff2bf988aa0edf2b34c02d220a45c4a3c572dac6b995771ed20de58da919bfa5*HfWzyJlz_UP9odmXvUaVK1TtdDuOCaezr-TAg2GjBCU"
      assert Iron.unseal(ticket, password) == {:error, {500, "Wrong mac prefix"}}
    end

    test "returns an error when integrity check fails", %{password: password} do
      ticket = "Fe26.2**b3ad22402ccc60fa4d527f7d1c9ff2e37e9b2e5723e9e2ffba39a489e9849609*QKCeXLs6Rp7f4LL56V7hBg*OvZEoAq_nGOpA1zae-fAtl7VNCNdhZhCqo-hWFCBeWuTTpSupJ7LxQqzSQBRAcgw**72018a21d3fac5c1608a0f9e461de0fcf17b2befe97855978c17a793faa01db1*Qj53DFE3GZd5yigt-mVl9lnp0VUoSjh5a5jgDmod1EZ"
      assert Iron.unseal(ticket, password) == {:error, {500, "Bad hmac value"}}
    end

    test "returns an error when decryption fails", %{password: password} do
      mac_base_string = "Fe26.2**a6dc6339e5ea5dfe7a135631cf3b7dcf47ea38246369d45767c928ea81781694*D3DLEoi-Hn3c972TPpZXqw*mCBhmhHhRKk9KtBjwu3h-1lx1MHKkgloQPKRkQZxpnDwYnFkb3RqdVTQRcuhGf4M??*"
      options = Map.put(Iron.defaults.integrity, :salt, "ff2bf988aa0edf2b34c02d220a45c4a3c572dac6b995771ed20de58da919bfa5")
      mac = Iron.hmac(password, options, mac_base_string)
      ticket = "#{mac_base_string}*#{options.salt}*#{mac.digest}"
      assert Iron.unseal(ticket, password) == {:error, {500, "Invalid character"}}
    end

    test "returns an error when iv base64 decoding fails", %{password: password} do
      mac_base_string = "Fe26.2**a6dc6339e5ea5dfe7a135631cf3b7dcf47ea38246369d45767c928ea81781694*D3DLEoi-Hn3c972TPpZXqw??*mCBhmhHhRKk9KtBjwu3h-1lx1MHKkgloQPKRkQZxpnDwYnFkb3RqdVTQRcuhGf4M*"
      options = Map.put(Iron.defaults.integrity, :salt, "ff2bf988aa0edf2b34c02d220a45c4a3c572dac6b995771ed20de58da919bfa5")
      mac = Iron.hmac(password, options, mac_base_string)
      ticket = "#{mac_base_string}*#{options.salt}*#{mac.digest}"
      assert Iron.unseal(ticket, password) == {:error, {500, "Invalid character"}}
    end

    test "returns an error when decrypted object is invalid", %{password: password} do
      bad_json = "{asdasd"
      %{encrypted: encrypted, key: %{iv: iv, salt: salt}} = Iron.encrypt(password, Iron.defaults.encryption, bad_json)
      mac_base_string = Iron.mac_prefix() <> "**" <> salt <> "*" <> Base.url_encode64(encrypted, padding: false) <> "*" <> Base.url_encode64(iv, padding: false) <> "*"
      %{digest: digest, salt: hmac_salt} = Iron.hmac(password, Iron.defaults.integrity, mac_base_string)
      ticket = "#{mac_base_string}*#{hmac_salt}*#{digest}"
      assert Iron.unseal(ticket, password) == {:error, {500, "Failed parsing sealed object JSON"}}
    end

    test "returns an error when expired", %{password: password} do
      mac_base_string = "Fe26.2**a38dc7a7bf2f8ff650b103d8c669d76ad219527fbfff3d98e3b30bbecbe9bd3b*nTsatb7AQE1t0uMXDx-2aw*uIO5bRFTwEBlPC1Nd_hfSkZfqxkxuY1EO2Be_jJPNQCqFNumRBjQAl8WIKBW1beF*1380495854060"
      integrity = Map.put(Iron.defaults.integrity, :salt, "e4fe33b6dc4c7ef5ad7907f015deb7b03723b03a54764aceeb2ab1235cc8dce3")
      mac = Iron.hmac(password, integrity, mac_base_string)
      ticket = "#{mac_base_string}*#{integrity.salt}*#{mac.digest}"
      assert Iron.unseal(ticket, password) == {:error, {500, "Expired seal"}}
    end

    test "returns an error when expiration NaN", %{password: password} do
      mac_base_string = "Fe26.2**a38dc7a7bf2f8ff650b103d8c669d76ad219527fbfff3d98e3b30bbecbe9bd3b*nTsatb7AQE1t0uMXDx-2aw*uIO5bRFTwEBlPC1Nd_hfSkZfqxkxuY1EO2Be_jJPNQCqFNumRBjQAl8WIKBW1beF*a"
      integrity = Map.put(Iron.defaults.integrity, :salt, "e4fe33b6dc4c7ef5ad7907f015deb7b03723b03a54764aceeb2ab1235cc8dce3")
      mac = Iron.hmac(password, integrity, mac_base_string)
      ticket = "#{mac_base_string}*#{integrity.salt}*#{mac.digest}"
      assert Iron.unseal(ticket, password) == {:error, {500, "Invalid expiration"}}
    end
  end
end


mix.exs
defmodule Iron.Mixfile do
  use Mix.Project

  @version "0.1.0"

  def project do
    [
      app: :iron,
      version: @version,
      elixir: "~> 1.5",
      name: "iron",
      source_url: "https://github.com/schultzer/iron",
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
      deps: deps(),
      description: description(),
      package: package(),
      docs: docs()
    ]
  end

  def application do
    [
      extra_applications: [:logger],
    ]
  end

  defp deps do
    [
      {:jason, "~> 1.1.1"},
      {:kryptiles, "~> 0.1.0"},
      {:ex_doc, "~> 0.14", only: :dev, runtime: false},
    ]
  end
  defp description do
    """
    Serialize and encrypt any Elixir `term()` and send it around.
    """
  end

  defp package do
    [
      name: :iron,
      maintainers: ["Benjamin Schultzer"],
      licenses: ~w(MIT),
      links: links(),
      files: ~w(CHANGELOG* README* lib mix.exs)
    ]
  end

  def docs do
    [
      source_ref: "v#{@version}",
      main: "readme",
      extras: ["README.md", "CHANGELOG.md"]
    ]
  end

  def links do
    %{
      "GitHub"    => "https://github.com/schultzer/iron",
      "Readme"    => "https://github.com/schultzer/iron/blob/v#{@version}/README.md",
      "Changelog" => "https://github.com/schultzer/iron/blob/v#{@version}/CHANGELOG.md"
    }
  end
end


lib/iron.ex
defmodule Iron do
  @moduledoc """
  Serialize and encrypt any Elixir `term()` and send it around.

  ## Examples

      iex> Iron.seal(%{"very" => "secret"}, "password")
      {:error, "password string too short (min 32 characters required)"}

      iex> sealed = Iron.seal(%{"a" => 1, "b" => 2, "c" => [3, 4, 5], "d" => %{ "e" => "f" }}, "some_not_random_password_that_is_also_long_enough")
      iex> Iron.unseal(sealed, "some_not_random_password_that_is_also_long_enough")
      %{"a" => 1, "b" => 2, "c" => [3, 4, 5], "d" => %{ "e" => "f" }}
  """
  use Bitwise

  @typedoc false
  @type password :: iodata() | %{id: iodata(), secret: iodata()} | %{id: iodata(), encryption: iodata(), integrity: iodata()}

  @typedoc """
  These are the Iron defualts it is recommended to use the defualts
  """
  @type defaults :: %{encryption: %{algorithm: :aes_cbc256, iterations: 1, iv_bits: 128, key_bits: 256, min_password_length: 32, salt_bits: 256},
                      integrity:  %{algorithm: :sha256, iterations: 1, key_bits: 256, min_password_length: 32, salt_bits: 256},
                      ttl: 0, timestamp_skew_sec: 60, localtime_offset_msec: 0}

  @doc false
  @spec algorithms() :: %{aes_ctr128: %{key_bits: 128, iv_bits: 128}, aes_cbc256: %{key_bits: 256, iv_bits: 128}, sha256: %{key_bits: 256}}
  def algorithms(), do: %{aes_ctr128: %{key_bits: 128, iv_bits: 128}, aes_cbc256: %{key_bits: 256, iv_bits: 128}, sha256: %{key_bits: 256}}

  # aes_ctr128 is supported if you want you use aes_ctr128 then you can use the map below
  # @aes_ctr128 %{iterations: 10000, iv_bits: 128, key_bits: 128, min_password_length: 32, salt_bits: 256}
  @aes_cbc256 %{algorithm: :aes_cbc256, iterations: 1, iv_bits: 128, key_bits: 256, min_password_length: 32, salt_bits: 256}
  @sha256 %{algorithm: :sha256, iterations: 1, key_bits: 256, min_password_length: 32, salt_bits: 256}

  @defaults %{encryption: @aes_cbc256, integrity: @sha256, ttl: 0, timestamp_skew_sec: 60, localtime_offset_msec: 0}
  @doc false
  def defaults(), do: @defaults

  @mac_format_version "2"
  @mac_prefix "Fe26." <> @mac_format_version

  @doc false
  def mac_prefix(), do: @mac_prefix

  @doc false
  @spec generate_key(binary(), keyword() | map()) :: map() | {:error, term()}
  def generate_key(password, options) when is_list(options), do: generate_key(password, Map.new(options))
  def generate_key(password, _options) when not is_binary(password) and not is_list(password), do: {:error, {500, "Invalid password #{inspect password}"}}
  def generate_key([], _options), do: {:error, {500, "Empty password"}}
  def generate_key(<<>>, _options), do: {:error, {500, "Empty password"}}
  def generate_key(_password, %{algorithm: algorithm}) when algorithm not in [:aes_ctr128, :aes_cbc256, :sha256] do
    {:error, {500, "Unknown algorithm: #{inspect algorithm}"}}
  end
  def generate_key(password, %{key_bits: key_bits, min_password_length: min_password_length} = options) when is_binary(password) do
    case String.valid?(password) do
      false when byte_size(password) < div(key_bits, 8)    -> {:error, {500, "Key buffer (password) too small"}}

      true  when byte_size(password) < min_password_length -> {:error, {500, "Password string too short (min #{min_password_length} characters required)"}}

      false ->  maybe_add_iv(%{key: password, salt: ""} , options)

      true  -> __generate_key__(password, options)
    end
  end
  def generate_key(password, %{min_password_length: min_password_length}) when is_list(password) and length(password) < min_password_length do
    {:error, {500, "Password too short (min #{min_password_length} characters required)"}}
  end
  def generate_key(_password, %{algorithm: _, iterations: _, key_bits: _, min_password_length: _}), do: {:error, {500, "Missing salt or salt_bits options"}}
  def generate_key(_password, _option), do: {:error, {500, "Bad options"}}

  defp __generate_key__(password, %{iterations: iterations, key_bits: key_bits, salt: salt} = options) do
    derived_key = Kryptiles.pbkdf2(password, salt, div(key_bits, 8), iterations)
    maybe_add_iv(%{key: derived_key, salt: salt}, options)
  end
  defp __generate_key__(password, %{iterations: iterations, key_bits: key_bits, salt_bits: salt_bits} = options) do
    case Kryptiles.random_bits(salt_bits) do
      {:error, reason} -> {:error, reason}

      bytes            ->
        salt = Base.hex_encode32(bytes)
        derived_key = Kryptiles.pbkdf2(password, salt, div(key_bits, 8), iterations)
        maybe_add_iv(%{key: derived_key, salt: salt}, options)
    end
  end
  defp __generate_key__(_password, %{algorithm: _, iterations: _, key_bits: _, min_password_length: _}), do: {:error, {500, "Missing salt or salt_bits options"}}
  defp __generate_key__(_password, _option), do: {:error, {500, "Bad options"}}

  defp maybe_add_iv(result, %{iv: iv}), do: Map.put(result, :iv, iv)
  defp maybe_add_iv(result, %{iv_bits: iv_bits}) do
    case Kryptiles.random_bits(iv_bits) do
      {:error, reason} -> {:error, reason}

      iv               -> Map.put(result, :iv, iv)
    end
  end
  defp maybe_add_iv(result, _), do: result

  @doc false
  @spec encrypt(binary(), map(), binary()) :: %{encrypted: binary(), key: map()} | {:error, binary()}
  def encrypt(password, %{algorithm: algorithm} = options, data) do
    case generate_key(password, options) do
      {:error, reason}              -> {:error, reason}

      %{iv: iv, key: gen_key} = key ->
      encrypted = :crypto.block_encrypt(algorithm, gen_key, iv, padding(data))
      %{encrypted: encrypted, key: key}
    end
  end

  @doc false
  @spec decrypt(binary(), map(), binary()) :: binary() | {:error, binary()}
  def decrypt(password, %{algorithm: algorithm} = options, data) do
    case generate_key(password, options) do
      {:error, reason}        -> {:error, reason}

      %{iv: iv, key: gen_key} -> :crypto.block_decrypt(algorithm, gen_key, iv, data)
      end
  end

  @doc false
  @spec hmac(binary(), map(), binary()) :: %{digest: binary(), salt: binary()} | {:error, binary()}
  def hmac(password, %{algorithm: algorithm} = options, data) do
    case generate_key(password, options) do
      {:error, reason}            -> {:error, reason}

      %{key: gen_key, salt: salt} ->
        hmac = :crypto.hmac(algorithm, gen_key, data)
        digest = hmac |> Base.encode64() |> __encode__()
        %{digest: digest, salt: salt}
    end
  end

  defp __encode__(binary, result \\ <<>>)
  defp __encode__(<<>>, result), do: result
  for {pattern, replacment} <- [{?+, ?-}, {?/, ?_}, {?=, <<>>}] do
    defp __encode__(<<unquote(pattern), rest::binary()>>, result), do: __encode__(rest, <<result::binary(), unquote(replacment)>>)
  end
  defp __encode__(<<b, rest::binary()>>, result), do: __encode__(rest, <<result::binary(), b>>)

  @doc """
  Encrypt and HMAC a `term()`
  """
  @spec seal(term(), password(), keyword() | map()) :: binary() | {:error, binary()}
  def seal(term, password, options \\ @defaults)
  def seal(term, password, options) when is_list(options), do: seal(term, password, Map.new(options))
  def seal(term, password, options) do
    now = System.system_time(1000) + (options[:localtime_offset_msec] || 0)
    term
    |> Jason.encode()
    |> __seal__(now, password, options)
  end

  defp __seal__({:error, reason}, _now, _password, _options), do: {:error, reason}
  defp __seal__({:ok, binary}, now, password, options) do
    case normalize_password(password) do
      {:error, reason} -> {:error, reason}

      %{id: id, encryption: encryption, integrity: integrity} ->
        case encrypt(encryption, options[:encryption], binary) do
          {:error, reason}                                    -> {:error, reason}

          %{encrypted: encrypted, key: %{iv: iv, salt: salt}} ->
            encrypted_b64 = Base.url_encode64(encrypted, padding: false)
            iv = Base.url_encode64(iv, padding: false)
            expiration = if options[:ttl], do: now + options[:ttl], else: ""
            mac_base_string = "#{@mac_prefix}*#{id}*#{salt}*#{iv}*#{encrypted_b64}*#{expiration}"
            case hmac(integrity, options[:integrity], mac_base_string) do
              {:error, reason}                  -> {:error, reason}

              %{salt: mac_salt, digest: digest} ->
              "#{mac_base_string}*#{mac_salt}*#{digest}"
            end
        end
    end
  end

  @doc """
  Decrypt and validate a sealed `binary()`
  """
  @spec unseal(binary(), password(), keyword() | map()) :: term() | {:error, binary()}
  def unseal(sealed, password, options \\ @defaults)
  def unseal(sealed, password, options) when is_list(options), do: unseal(sealed, password, Map.new(options))
  def unseal(sealed, password, options) do
    now = System.system_time(1000) + (options[:localtime_offset_msec] || 0)
    case :string.split(sealed, '*', :all) do
      parts when length(parts) !== 8                   -> {:error, {500, "Incorrect number of sealed components"}}

      [mac_prefix | _] when mac_prefix !== @mac_prefix -> {:error, {500, "Wrong mac prefix"}}

      parts ->
        [:mac_prefix, :password_id, :encryption_salt, :encryption_iv, :encrypted_b64, :expiration, :hmac_salt, :hmac]
        |> Enum.zip(parts)
        |> Enum.into(%{})
        |> maybe_validate_expiration(now, options)
        |> __unseal__(password, options)
      end
  end

  defp maybe_validate_expiration(%{expiration: ""} = parts, _now, _options), do: parts
  defp maybe_validate_expiration(%{expiration: expiration} = parts, now, options) do
    skew = options[:timestamp_skew_sec]
    case :string.to_integer(expiration) do
      {exp, ""} when exp <= (now - (skew * 1000)) -> {:error, {500, "Expired seal"}}

      {_exp, ""}                                  -> parts

      {_, _}                                      -> {:error, {500, "Invalid expiration"}}
    end
  end

  defp __unseal__({:error, reason}, _password_or_parts, _options), do: {:error, reason}
  defp __unseal__(%{id: id}, %{password_id: password_id}, _options) when id !== password_id do
    {:error, {500, "Cannot find password: #{password_id}"}}
  end
  defp __unseal__(%{encryption: encryption, integrity: integrity}, %{mac_prefix: mac_prefix, password_id: password_id, encryption_salt: encryption_salt, encryption_iv: encryption_iv, encrypted_b64: encrypted_b64, expiration: expiration, hmac_salt: hmac_salt} = parts, options) do
    mac_base_string = <<mac_prefix::binary(), ?*, password_id::binary(), ?*, encryption_salt::binary(), ?*, encryption_iv::binary(), ?*, encrypted_b64::binary(), ?*, expiration::binary()>>
    hmac_opts = Map.put(options[:integrity], :salt, hmac_salt)

    integrity
    |> hmac(hmac_opts, mac_base_string)
    |> __unseal__(encryption, parts, options)
  end
  defp __unseal__(parts, password, options) do
    password
    |> normalize_password()
    |> __unseal__(parts, options)
  end
  defp __unseal__(%{digest: digest}, encryption, %{hmac: hmac} = parts, options) do
    digest
    |> Kryptiles.fixed_time_comparison(hmac)
    |> maybe_decrypt(encryption, parts, options)
  end
  defp __unseal__(error, _encryption, _parts, _options), do: error

  defp maybe_decrypt(false, _encryption, _parts, _options), do: {:error, {500, "Bad hmac value"}}
  defp maybe_decrypt(true, encryption, %{encryption_salt: encryption_salt, encryption_iv: encryption_iv, encrypted_b64: encrypted_b64}, options) do
    case url_decode64(encrypted_b64, encryption_iv) do
      {:error, reason}           -> {:error, reason}

      {enc, iv} ->
        decrypt_opts = Map.merge(options[:encryption], %{salt: encryption_salt, iv: iv})

        encryption
        |> decrypt(decrypt_opts, enc)
        |> unpadding()
        |> decode_json()
    end
  end

  def decode_json(:error), do: {:error, {500, "Failed parsing sealed object JSON"}}
  def decode_json(msg) do
    case Jason.decode(msg, keys: &mixed_keys/1) do
      {:error, _} -> {:error, {500, "Failed parsing sealed object JSON"}}

      {:ok, map}  -> map
    end
  end

  defp mixed_keys(key) do
    try do
      String.to_existing_atom(key)
    rescue
      ArgumentError -> key
    end
  end

  defp url_decode64({:ok, encrypted}, {:ok, encryption_iv}), do: {encrypted, encryption_iv}
  defp url_decode64(enc, iv) when is_binary(enc) and is_binary(iv) do
    enc
    |> Base.url_decode64(padding: false)
    |> url_decode64(Base.url_decode64(iv, padding: false))
  end
  defp url_decode64(_, _), do: {:error, {500, "Invalid character"}}

  defp normalize_password(%{id: id, encryption: _, integrity: _} = password) do
    case String.match?(id, ~r/^\w+$/) do
      false -> {:error, {500, "Invalid password id: #{inspect id}"}}

      true  -> password
    end

  end
  defp normalize_password(%{id: id, secret: secret}) do
    case String.match?(id, ~r/^\w+$/) do
      false -> {:error, {500, "Invalid password id: #{inspect id}"}}

      true  -> %{id: id, encryption: secret, integrity: secret}
    end
  end
  defp normalize_password(""), do: {:error, {500, "Empty password"}}
  defp normalize_password(''), do: {:error, {500, "Empty password"}}
  defp normalize_password(password) when is_binary(password) or is_list(password), do: %{id: <<>>, encryption: password, integrity: password}
  defp normalize_password(password), do: {:error, {500, "Invalid password #{inspect password}"}}

  # https://tools.ietf.org/html/rfc2315
  defp padding(message) do
    padding_size = 16 - rem(byte_size(message), 16)
    message <> :binary.copy(<<padding_size>>, padding_size)
  end
  defp unpadding(<<>>), do: :error
  defp unpadding(message) do
    padding_size = :binary.last(message)
    padding = byte_size(message) - padding_size
    case padding_size <= 16 do
      true  -> :erlang.binary_part(message, 0, padding)

      false -> :error
    end
  end
end

